Import các thư viện cần dùng:

In [69]:
import numpy as np
from datascience import * 

import matplotlib
%matplotlib inline
import matplotlib.pyplot as plt
plt.style.use('fivethirtyeight')
import warnings
warnings.simplefilter('ignore', FutureWarning)
from datetime import datetime

Import tập dữ liệu

In [70]:
student_peformance_table = Table().read_table('student_performance_large_dataset.csv')
student_peformance_table.show(5)

Phân tích các đặc điểm cơ bản của tập dữ liệu:

Bảng có 10000 hàng và 15 cột:

In [71]:
student_peformance_table.num_rows, student_peformance_table.num_columns

(10000, 15)

Tên của các cột:

In [72]:
student_peformance_table.labels

('Student_ID',
 'Age',
 'Gender',
 'Study_Hours_per_Week',
 'Preferred_Learning_Style',
 'Online_Courses_Completed',
 'Participation_in_Discussions',
 'Assignment_Completion_Rate (%)',
 'Exam_Score (%)',
 'Attendance_Rate (%)',
 'Use_of_Educational_Tech',
 'Self_Reported_Stress_Level',
 'Time_Spent_on_Social_Media (hours/week)',
 'Sleep_Hours_per_Night',
 'Final_Grade')

Kiểm tra xem có giá trị null không:

In [73]:
def has_missing(array):
  '''Hàm kiểm tra xem 1 mảng có chứa giá trị null không'''
  for item in array:
      if item is None:
          return True
  return False

is_missing = make_array()
for col in student_peformance_table.labels:
  is_missing = np.append(is_missing, has_missing(col))

is_missing

array([ 0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,
        0.,  0.])

Vậy, không có giá trị null nào trong bảng.

# Câu hỏi nghiên cứu 1:
Học sinh có điểm thi (Exam_Score) dưới 60% và tỷ lệ hoàn thành bài tập dưới 70%(Assignment_Completion_Rate < 70%) có xu hướng đạt điểm cuối kỳ D không?

Đầu tiên, ta chọn những cột thông tin cần thiết:

In [74]:
student_performance_table = Table().read_table('student_performance_large_dataset.csv')
student_performance_table

Student_ID | Age  | Gender | Study_Hours_per_Week | Preferred_Learning_Style | Online_Courses_Completed | Participation_in_Discussions | Assignment_Completion_Rate (%) | Exam_Score (%) | Attendance_Rate (%) | Use_of_Educational_Tech | Self_Reported_Stress_Level | Time_Spent_on_Social_Media (hours/week) | Sleep_Hours_per_Night | Final_Grade
S00001     | 18   | Female | 48                   | Kinesthetic              | 14                       | Yes                          | 100                            | 69             | 66                  | Yes                     | High                       | 9                                       | 8                     | C
S00002     | 29   | Female | 30                   | Reading/Writing          | 20                       | No                           | 71                             | 40             | 57                  | Yes                     | Medium                     | 28                                      | 8                     | D
S00003     | 20   | Female | 47                   | Kinesthetic              | 11                       | No                           | 60                             | 43             | 79                  | Yes                     | Low                        | 13                                      | 7                     | D
S00004     | 23   | Female | 13                   | Auditory                 | 0                        | Yes                          | 63                             | 70             | 60                  | Yes                     | Low                        | 24                                      | 10                    | B
S00005     | 19   | Female | 24                   | Auditory                 | 19                       | Yes                          | 59                             | 63             | 93                  | Yes                     | Medium                     | 26                                      | 8                     | C
S00006     | 28   | Female | 26                   | Kinesthetic              | 5                        | Yes                          | 63                             | 54             | 80                  | No                      | High                       | 25                                      | 8                     | D
S00007     | 19   | Female | 49                   | Reading/Writing          | 13                       | Yes                          | 91                             | 44             | 66                  | Yes                     | Low                        | 30                                      | 10                    | D
S00008     | 27   | Male   | 14                   | Reading/Writing          | 5                        | Yes                          | 88                             | 56             | 76                  | Yes                     | Low                        | 4                                       | 6                     | C
S00009     | 22   | Male   | 45                   | Visual                   | 16                       | No                           | 52                             | 78             | 70                  | No                      | Low                        | 26                                      | 9                     | B
S00010     | 28   | Other  | 35                   | Auditory                 | 7                        | No                           | 100                            | 55             | 100                 | No                      | Medium                     | 5                                       | 9                     | C
... (9990 rows omitted)

# Lọc nhóm học sinh có Exam_Score < 60% và Assignment_Completion_Rate < 70%

In [75]:
filtered_table = student_performance_table.where('Exam_Score (%)', are.below(60)).where('Assignment_Completion_Rate (%)', are.below(70))
filtered_table

Student_ID | Age  | Gender | Study_Hours_per_Week | Preferred_Learning_Style | Online_Courses_Completed | Participation_in_Discussions | Assignment_Completion_Rate (%) | Exam_Score (%) | Attendance_Rate (%) | Use_of_Educational_Tech | Self_Reported_Stress_Level | Time_Spent_on_Social_Media (hours/week) | Sleep_Hours_per_Night | Final_Grade
S00003     | 20   | Female | 47                   | Kinesthetic              | 11                       | No                           | 60                             | 43             | 79                  | Yes                     | Low                        | 13                                      | 7                     | D
S00006     | 28   | Female | 26                   | Kinesthetic              | 5                        | Yes                          | 63                             | 54             | 80                  | No                      | High                       | 25                                      | 8                     | D
S00031     | 20   | Male   | 23                   | Kinesthetic              | 3                        | Yes                          | 59                             | 57             | 81                  | Yes                     | Medium                     | 9                                       | 5                     | C
S00035     | 28   | Female | 32                   | Auditory                 | 5                        | No                           | 53                             | 54             | 96                  | Yes                     | High                       | 0                                       | 10                    | D
S00037     | 26   | Female | 26                   | Kinesthetic              | 3                        | Yes                          | 53                             | 48             | 76                  | No                      | Low                        | 19                                      | 4                     | D
S00041     | 27   | Male   | 41                   | Reading/Writing          | 14                       | No                           | 65                             | 56             | 74                  | Yes                     | Low                        | 26                                      | 9                     | C
S00054     | 19   | Male   | 47                   | Kinesthetic              | 2                        | No                           | 52                             | 44             | 84                  | Yes                     | Medium                     | 11                                      | 10                    | D
S00059     | 29   | Female | 36                   | Reading/Writing          | 17                       | Yes                          | 52                             | 46             | 51                  | Yes                     | Low                        | 27                                      | 7                     | D
S00067     | 24   | Male   | 12                   | Kinesthetic              | 14                       | No                           | 57                             | 57             | 73                  | Yes                     | Medium                     | 4                                       | 7                     | C
S00072     | 20   | Female | 46                   | Visual                   | 13                       | No                           | 55                             | 45             | 52                  | No                      | Medium                     | 7                                       | 7                     | D
... (1261 rows omitted)

Tiếp thep, ta tính tỉ lệ phần trăm học sinh trong nhóm này bị điểm D:


In [76]:
num_filtered = filtered_table.num_rows
num_D_filtered = np.count_nonzero(filtered_table.column('Final_Grade') == 'D')
percentage_D_filtered = (num_D_filtered / num_filtered * 100) if num_filtered > 0 else 0
percentage_D_filtered

75.68843430369787

Sau đó, ta tính tỉ lệ phần trăm học sinh bị điểm D trong cả tập dữ liệu rồi so sánh hai số liệu này với nhau:

In [77]:
num_total = student_performance_table.num_rows
num_D_total = np.count_nonzero(student_performance_table.column('Final_Grade') == 'D')
percentage_D_population = (num_D_total / num_total) * 100
percentage_D_population

24.27

Lọc nhóm học sinh có Exam_Score (%) từ 60% trở lên và Assignment_Completion_Rate (%) từ 70% trở lên để so sánh:

In [78]:
condition = (student_performance_table.column('Exam_Score (%)') >= 60) | (student_performance_table.column('Assignment_Completion_Rate (%)') >= 70)
opposite_table = student_performance_table.where(condition)
num_opposite = opposite_table.num_rows
num_D_opposite = np.count_nonzero(opposite_table.column('Final_Grade') == 'D')
percentage_D_opposite = (num_D_opposite / num_opposite * 100) if num_opposite > 0 else 0
percentage_D_opposite

16.783136670867226

Học sinh có điểm thi dưới 60% và tỷ lệ hoàn thành bài tập dưới 70% có xu hướng mạnh mẽ đạt điểm cuối kỳ D, với tỷ lệ 75.69%, cao gấp hơn 4.5 lần so với nhóm đối lập (16.8%) và gấp hơn 3 lần so với trung bình toàn bộ tập dữ liệu (24.27%). Điều này cho thấy các yếu tố điểm thi thấp và tỷ lệ hoàn thành bài tập thấp là những chỉ số mạnh mẽ dự đoán điểm cuối kỳ D. Để cải thiện kết quả học tập, cần can thiệp hỗ trợ học sinh trong việc nâng cao điểm thi và hoàn thành bài tập đúng hạn.

# Câu hỏi nghiên cứu 2:
Học sinh có phong cách học tập ưa thích là 'Visual' (Preferred_Learning_Style = Visual) và dành ít hơn 15 giờ học mỗi tuần (Study_Hours_per_Week < 15) có xu hướng đạt điểm cuối kỳ dưới B không?


Ta lấy từ dataset ra những cột cần thiết:

In [79]:
table_for_q2 = student_peformance_table.select('Preferred_Learning_Style', 'Study_Hours_per_Week', 'Final_Grade')
table_for_q2

Preferred_Learning_Style | Study_Hours_per_Week | Final_Grade
Kinesthetic              | 48                   | C
Reading/Writing          | 30                   | D
Kinesthetic              | 47                   | D
Auditory                 | 13                   | B
Auditory                 | 24                   | C
Kinesthetic              | 26                   | D
Reading/Writing          | 49                   | D
Reading/Writing          | 14                   | C
Visual                   | 45                   | B
Auditory                 | 35                   | C
... (9990 rows omitted)

Lọc nhóm học sinh có phong cách học 'Visual' và học dưới 15 giờ/tuần

In [80]:
filtered_table_q2 = table_for_q2.where('Preferred_Learning_Style', are.equal_to('Visual')).where('Study_Hours_per_Week', are.below(15))
filtered_table_q2

Preferred_Learning_Style | Study_Hours_per_Week | Final_Grade
Visual                   | 14                   | C
Visual                   | 11                   | A
Visual                   | 6                    | A
Visual                   | 11                   | B
Visual                   | 12                   | C
Visual                   | 6                    | A
Visual                   | 5                    | C
Visual                   | 11                   | B
Visual                   | 9                    | B
Visual                   | 12                   | B
... (573 rows omitted)

Tính tỷ lệ học sinh đạt điểm dưới B trong nhóm đã lọc

In [81]:
percentage_below_B_filtered = np.count_nonzero(np.isin(filtered_table_q2.column('Final_Grade'), ['C', 'D'])) / filtered_table_q2.num_rows * 100
percentage_below_B_filtered

49.74271012006861

So sánh với nhóm đối lập

In [82]:
opposite_group = table_for_q2.where('Preferred_Learning_Style', are.not_equal_to('Visual')).where('Study_Hours_per_Week', are.above_or_equal_to(15))
percentage_below_B_opposite = np.count_nonzero(np.isin(opposite_group.column('Final_Grade'), ['C', 'D'])) / opposite_group.num_rows * 100
percentage_below_B_opposite

48.832054560954816

Học sinh có phong cách học tập 'Visual' và học dưới 15 giờ/tuần có xu hướng đạt điểm cuối kỳ dưới B (C hoặc D) với tỷ lệ 53.81%, cao hơn so với toàn bộ tập dữ liệu (49.74%) và nhóm không phải Visual hoặc học từ 15 giờ/tuần (48.83%).